doc/neps/nep-0019-rng-policy.rst

```diff
@@ -0,0 +1,208 @@
+a significant increase in performance, for example, moving from the `Box-Muller
+transform <https://en.wikipedia.org/wiki/Box%E2%80%93Muller_transform>`_ method
+of Gaussian variate generation to the faster `Ziggurat algorithm
+<https://en.wikipedia.org/wiki/Ziggurat_algorithm>`_.  An example of an
+unworthy improvement would be tweaking the Ziggurat tables just a little bit.
+
+Any new design for the RNG subsystem will provide a choice of different core
+uniform PRNG algorithms.  We will be more strict about a select subset of
+methods on these core PRNG objects.  They MUST guarantee stream-compatibility
+for a minimal, specified set of methods which are chosen to make it easier to
+compose them to build other distributions.  Namely,
+
+    * ``.bytes()``
+    * ``.random_uintegers()``
+    * ``.random_sample()``
+
+Furthermore, the new design should also provide one generator class (we shall
+call it ``StableRandom`` for discussion purposes) that provides a slightly
+broader subset of distribution methods for which stream-compatibility is
+*guaranteed*.  The point of ``StableRandom`` is to provide something that can
+be used in unit tests so projects that currently have tests which rely on the
+precise stream can be migrated off of ``RandomState``.  For the best
+transition, ``StableRandom`` should use as its core uniform PRNG the current
+MT19937 algorithm.  As best as possible, the API for the distribution methods
+that are provided on ``StableRandom`` should match their counterparts on
+``RandomState``.  They should provide the same stream that the current version
+of ``RandomState`` does.  Because their intended use is for unit tests, we do
+not need the performance improvements from the new algorithms that will be
+introduced by the new subsystem.
+
+The list of ``StableRandom`` methods should be chosen to support unit tests:
+
+    * ``.randint()``
+    * ``.uniform()``
+    * ``.normal()``
+    * ``.standard_normal()``
+    * ``.choice()``
+    * ``.shuffle()``
+    * ``.permutation()``
+
+
+Not Versioning
+--------------
+
+For a long time, we considered that the way to allow algorithmic improvements
+while maintaining the stream was to apply some form of versioning.  That is,
+every time we make a stream change in one of the distributions, we increment
+some version number somewhere.  ``numpy.random`` would keep all past versions
+of the code, and there would be a way to get the old versions.  Proposals of
+how to do this exactly varied widely, but we will not exhaustively list them
+here.  We spent years going back and forth on these designs and were not able
+to find one that sufficed.  Let that time lost, and more importantly, the
+contributors that we lost while we dithered, serve as evidence against the
+notion.
+
+Concretely, adding in versioning makes maintenance of ``numpy.random``
+difficult.  Necessarily, we would be keeping lots of versions of the same code
+around.  Adding a new algorithm safely would still be quite hard.
```

Comment:
I agree with this.  It is especially hard since internally many generators depend on other C-implementations of generators (e.g., Student's t on Normal), which makes versioned Student's T harder.